### PR TITLE
Automated Transfers v1

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR adds the ability to install plugins for wp.com (non-atomic) business plan sites. Trying to install a plugin will start an Automated Transfer. Its steps are explained as comments in `PluginDetailActivity`. The changes to FluxC are added in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/775. Although it'd be good to check out the description of that PR, it's not necessary to review this one or to merge it since it's targeting a WIP branch.

To test:
1. Create a new business plan site on WordPress.com
2. Go into the plugins page for that site
3. Tap on any plugin and try to install it
4. Verify that you get an error
5. Add a custom domain to your site
6. Try to install the plugin again, make sure the progress bar is updated and the view refreshes after it's complete (the install button should say `Installed` for example)
7. Go back to the plugin directory page and make sure the site plugins showed up

### CODE REVIEW

**Comment by nbradbury** [comment]
This dialog is pretty wordy. Maybe "Installing the first plugin on your site can take a minute. You won't be able to access your site until it's done. Continue?" would work?

screenshot_1523294499

**Comment by nbradbury** [comment]
I got this error when I tried to install a plugin on a private site. 

screenshot_1523294528

File: WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java [hunk 1]
```diff
@@ -87,4 +88,12 @@ public static boolean checkMinimalJetpackVersion(SiteModel site, String limitVer
         }
         return false;
     }
+
+    public static boolean isNonAtomicBusinessPlanSite(@Nullable SiteModel site) {
+        return site != null && !site.isAutomatedTransfer() && SiteUtils.hasBusinessPlan(site);
+    }
+
+    public static boolean hasBusinessPlan(SiteModel site) {
+        return site.getPlanShortName() != null && site.getPlanShortName().equalsIgnoreCase("business");
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java]: Can we avoid hard-coding "business" here? Maybe this would work:

`return site.getPlanId() == PlansConstants.BUSINESS_PLAN_ID
               || site.getPlanId() == PlansConstants.JETPACK_BUSINESS_PLAN_ID;`

**Comment by nbradbury** [comment]
I got the first error when I tried to install a plugin on a private site, and the second error when I tried to install on a site that didn't have a custom domain. Can we handle these situations better, before the user tries to install?

screenshot_1523294528

screenshot_1523295104

**Comment by nbradbury** [comment]
I noticed that I can install Jetpack on my wp.com site. Should we be hiding Jetpack?

screenshot_1523296166

**Comment by nbradbury** [comment]
I tested installing a plugin on a slow connection and it appeared to be working, but then the progress stopped at 99%. I waited a full five minutes but the dialog never went away or got to 100%. I didn't see anything relevant in the log.

Aside: Should the dialog title be "Installing Plugin"?

screenshot_1523296728

**Comment by oguzkocer** [comment]
@nbradbury Thanks a lot for the review. Here are my combined notes to your comments:

> This dialog is pretty wordy. Maybe "Installing the first plugin on your site can take a minute. You won't be able to access your site until it's done. Continue?" would work?

I don't think it would be enough, as we need to tell them that if they make any changes that'll be lost, so they don't just force close the app and restart. We discussed this out in a p2 post and @SylvesterWilmott came up with this message.

> I got the first error when I tried to install a plugin on a private site, and the second error when I tried to install on a site that didn't have a custom domain (as mentioned in your description). Can we handle these situations better, before the user tries to install?

Thanks a lot for testing all these cases. I could swear that I wrote the error handling will happen in a different PR, but I can't find it. You can ignore these for now.

> I noticed that I can install Jetpack on my wp.com site. Should we be hiding Jetpack?

They should be coming from the API as installed because it's part of the Automated Transfer flow in the back end, I'll check out why that's not the case. We do have another issue for showing the "Automanaged" label. #7367

> I tested installing a plugin on a slow connection and it appeared to be working, but then the progress stopped at 99%. I waited a full five minutes but the dialog never went away or got to 100%.

That's one of the problems with the Automated Transfer, unless we fetch the site after it's completed, you can't use the site, at all. It basically becomes a different site, since it'll be transferred away and the current information for the site will be invalidated. My guess is that the flow completed, but the fetch site never happened, that's why you got 99%. I'll see if we can improve on that, but that's not a blocker for this PR as it's being merged to WIP branch.

> Aside: Should the dialog title be "Installing Plugin"?

@SylvesterWilmott ^^

> On the above site with the authorization problem, I attempted to install the same plugin that looked like it failed and very quickly saw an "Error installing" snackbar. Looking at the log, the error was PLUGIN_ALREADY_INSTALLED. Is this something we can catch and handle better?

I am guessing this happened because the flow didn't fully complete as otherwise the plugin should have showed up as installed. I'll check out what we can do about that, worst case scenario I guess we can refetch the plugins to at least refresh the page.

**Comment by nbradbury** [comment]
As discussed by DM, looks like we need to update `PluginUtils.isPluginFeatureAvailable()` to return `false` for private sites.

screen shot 2018-04-09 at 4 46 35 pm

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java [hunk 2]
```diff
@@ -1066,4 +1113,238 @@ private boolean isPluginStateChangedSinceLastConfigurationDispatch() {
...
+                // and its plugins
+                mAutomatedTransferProgressDialog.setProgress(99);
+                mDispatcher.dispatch(SiteActionBuilder.newFetchSiteAction(mSite));
+            } else {
+                mAutomatedTransferProgressDialog.setProgress(event.currentStep * 100 / event.totalSteps);
+                mHandler.postDelayed(new Runnable() {
+                    @Override
+                    public void run() {
+                        // Let's wait a second before checking the status again
+                        mDispatcher.dispatch(SiteActionBuilder.newCheckAutomatedTransferStatusAction(mSite));
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: I'm wondering if we really need to recheck so soon - seems like we could recheck in a few seconds and still provide a good experience.

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java [hunk 3]
```diff
@@ -1066,4 +1113,238 @@ private boolean isPluginStateChangedSinceLastConfigurationDispatch() {
...
+        if (isFinishing()) {
+            return;
+        }
+        if (event.isError()) {
+            handleAutomatedTransferFailed(event.error.message);
+        } else {
+            if (event.isCompleted) {
+                // The flow is almost complete, we can show 99% complete to give us a second or so to fetch the site
+                // and its plugins
+                mAutomatedTransferProgressDialog.setProgress(99);
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: This is where the flow failed for me earlier and got stuck at 99%. Would it be better to change `mAutomatedTransferProgressDialog` to indeterminate here, and somehow change the dialog text to reflect that things are being wrapped up? That would've made me "feel" better when things went wrong, especially after seeing a warning that I couldn't change my site until this completed.

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java [hunk 4]
```diff
@@ -1066,4 +1113,238 @@ private boolean isPluginStateChangedSinceLastConfigurationDispatch() {
...
+     * will need to be updated because of it, but they are both still possible and we try to handle it with a refresh.
+     */
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onPluginDirectoryFetched(OnPluginDirectoryFetched event) {
+        // We should be safe to ignore any errors related to plugin directory fetches since the only time we trigger it
+        // from this view is during AT progress and it should hopefully always succeed and we can't really do much about
+        // it anyway.
+        if (isFinishing() || event.isError()) {
+            return;
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: If there's an error, shouldn't we hide the AT progress dialog? Otherwise it could be left showing and the user can't do anything to get rid of it (outside of killing the app).

Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: Yes, I am planning to do that, however I wasn't sure if it could have somehow killed the current AT process and wanted to make sure I handled it. I should have left a todo for that.

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java (hunk 3 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: I really like this idea, will implement it in this PR!

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java (hunk 2 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: We can definitely increase the duration for check, I just threw in some number there. I guess something in between 3 to 5 seconds could work?

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java [hunk 5]
```diff
@@ -1066,4 +1113,238 @@ private boolean isPluginStateChangedSinceLastConfigurationDispatch() {
...
+     */
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onAutomatedTransferEligibilityChecked(OnAutomatedTransferEligibilityChecked event) {
+        if (isFinishing()) {
+            return;
+        }
+        if (!event.isEligible) {
+            String message =
+                    event.eligibilityErrors.isEmpty() ? getString(R.string.plugin_install_error_site_ineligible)
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: I noticed while testing the related FluxC branch, `onAutomatedTransferEligibilityChecked()` showed an error for that site of mine that I added a domain to yesterday. I think we can ignore that error - it's a result of the domain process, I'm sure - but note that `event.eligibilityErrors` is null in this error situation, so the line `event.eligibilityErrors.isEmpty()` will result in a NPE.

I think it's worth checking for null here, and also fixing FluxC because `eligibilityErrors` is supposed to be `@NonNull`.

screen shot 2018-04-10 at 9 25 45 am

Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: > but note that `event.eligibilityErrors` is `null` in this error situation

That was an oversight and was fixed in the FluxC branch. It's now an empty list in case of an error. We still need to handle that case a bit better here though, which I'll try to do so :)

File: WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java (hunk 1 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java]: Fixed in 11f057965f9382034b93566609a0c52cc7443f2a. I've decided to rename it and only return the non-jetpack business plans as the Jetpack ones should already have plugins feature available and it makes it easier to understand when Automated Transfer should be available. Let me know if you want me to change that back.

File: WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java (hunk 2 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java]: Fixed in e2335ba21ef1a1b1ed8b9fa4cb06abfd3c69574a.

**Comment by oguzkocer** [comment]
@nbradbury I've made the requested changes for using the plans constant, increasing the wait duration to check the AT status and hiding the plugins section for private sites as well as sites without a custom domain. I've created a checklist in #7653 for the rest of the issues as it has been really hard to follow the feedback in this PR. I'll tackle them in a separate PR to make it easier for the both of us.

**Comment by nbradbury** [comment]
@oguzkocer I just tested this with a second site, and just like the first it hung at 99%. I had to force quit the app.

screenshot_1524166587

There's obviously something wrong, but I think it's best for me to merge this into the feature branch and then you resolve that issue separately.

**Comment by oguzkocer** [comment]
> There's obviously something wrong, but I think it's best for me to merge this into the feature branch and then you resolve that issue separately.

@nbradbury I already have an item for that in the checklist.

**Comment by nbradbury** [comment]
:shipit:
